draft-ietf-emu-rfc7170bis.md

```diff
@@ -2840,7 +2840,7 @@ IMSK[j] = First 32 octets of TLS-PRF(secret, "TEAPbindkey@ietf.org",
 >
 > PRF(secret, label, seed) = P_\<hash>(secret, label \| seed).
 >
-> The secret is the EMSK or MSK from the j'th inner method, the label is
+> The secret is the EMSK from the j'th inner method, the label is
```

Comment:
I would probably change this to

```
Where the EMSK is available, the secret is the EMSK from ...
```

that mirrors the text in the next paragraph.

Reply:
Changed with amended commit and force push. Now it's clear that EMSK is not always available.

Reply:
Maybe I am reading too much into it, but here I see the *previous* paragraph reads "If an inner method supports export of an Extended Master Session Key..." so I am lead into this section as "skip this when not possible" but then the next bit "Where the EMSK is *available* [sic set the 'secret' to it]..." which I might interpret as bumping me back into "the secret is optional but not the calculation" so I may decide to use all NULLs as the secret similar to how the MSK fallback does (deduction via inferring symmetry where there is none).

There is a lot to unpack in this section, not much we can do about that, but I think we can help hugely by simply replacing `secret` with `EMSK[j]` in:

```
IMSK[j] = First 32 octets of TLS-PRF(EMSK[j], "TEAPbindkey@ietf.org",
       0x00 \| 0x00 \| 0x40)
```

Then we spoon feed below with "`EMSK[0]` = first 64 bytes of `session_key_seed` defined below"; this is in keeping with elsewhere when the draft talks of `S-IMCK[0]` and will not look out of place.

We should though *retain* unchanged the PRF function as the parameters *names* are useful to pair up with whatever library they may be using

```
PRF(secret, label, seed) = P_<hash>(secret, label | seed).
```

If someone actually reads this and does not skim this, it is very clear the pairing between what makes up the `IMSK[j]` and how to plug it into the `PRF`.

Reply:
@jimdigriz The secret can be `EMSK` or `MSK`. So I think it's too confusing to replace `secret` with `EMSK...`

Reply:
@alandekok Not for `TLS-PRF(secret, "TEAPbindkey@ietf.org", ...)` it isn't, that is only `EMSK` where the output is then fed into the `S-IMCK` PRF.

Meanwhile for the `MSK path, it is just fed verbatim directly into the `S-IMCK` PRF.

https://github.com/FreeRADIUS/freeradius-server/blob/v3.2.x/src/modules/rlm_eap/types/rlm_eap_teap/eap_teap.c#L89-L160

Right?

Reply:
...this said, @h-vn was able to implement TEAP from scratch using this document, so that is validation we are good enough.

Reply:
What we did was based on the RFC 7170 first and then upgraded when the draft progressed. I noticed the IMSK calculation change when I was reading through the diff between RFC 7170 and draft -13. At this point I noticed that they had diverged. In other words, the updates to the draft that touched IMSK calculation from EMSK and non-zero MSK were thought of as editorial changes only, because functional changes to these were not expected.

What's confusing about IMSK construction is that the EMSK ends up being passed to TLS-PRF, but the respective transformation for the MSK isn't done, or it can be considered a null operation. And when the IMSK is used, it's passed to the same TLS-PRF function again when `S-IMCK` is calculated.

Reply:
> sure? Do you have updated text you can suggest?

quite right, https://github.com/emu-wg/rfc7170bis/pull/28

Reply:
> What's confusing about IMSK construction is that the EMSK ends up being passed to TLS-PRF, but the respective transformation for the MSK isn't done, or it can be considered a null operation. And when the IMSK is used, it's passed to the same TLS-PRF function again when S-IMCK is calculated.

It is strange that the MSK path involves less work and is missing the symmetry of the EMSK path. I suspect this is a left over from [EAP-FAST](https://datatracker.ietf.org/doc/html/rfc4851#section-5.2) which also just eats the MSK directly, but whilst forming TEAP those implementing it want to retain that code but for the EMSK could not reconcile it with RFC5295 without the additional PRF step.

Reply:
The indirect EMSK use could come from this requirement:
https://www.rfc-editor.org/rfc/rfc5295.html#section-5

> o  The usage definition MUST NOT use the EMSK in any other way except
>       to derive Root Keys using the key derivation specified in
>       [Section 3](https://www.rfc-editor.org/rfc/rfc5295.html#section-3) of this document.  They MUST NOT use the EMSK directly.

I just noticed this when I took yet another look at RFC 5295.

Reply:
Do we need to update the text based on these comments?

Reply:
> Do we need to update the text based on these comments?

I don't think so, just banter, right?